api/python/tests/test_bucket.py

```diff
@@ -19,16 +17,98 @@ def test_bucket_construct(self):
         Bucket('s3://test-bucket')
 
     def test_bucket_fetch(self):
-        response = {
-            'IsTruncated': False
-        }
-        params = {
-            'Bucket': 'test-bucket',
-            'Prefix': 'does/not/exist/'
-        }
-        self.s3_stubber.add_response('list_objects_v2', response, params)
+        bucket = Bucket('s3://test-bucket')
+
+        a_contents = b'a' * 10
+        b_contents = b'b' * 20
+
+        # Fetch a directory.
+
+        self.s3_stubber.add_response(
+            method='list_objects_v2',
+            service_response={
+                'IsTruncated': False,
+                'Contents': [
+                    {'Key': 'dir/a', 'Size': len(a_contents)},
+                    {'Key': 'dir/foo/b', 'Size': len(b_contents)}
+                ],
+            },
+            expected_params={
+                'Bucket': 'test-bucket',
+                'Prefix': 'dir/'
+            }
+        )
+        self.s3_stubber.add_response(
+            method='get_object',
+            service_response={
+                'ContentLength': len(a_contents),
+                'Body': BytesIO(a_contents)
+            },
+            expected_params={
+                'Bucket': 'test-bucket',
+                'Key': 'dir/a'
+            }
+        )
+        self.s3_stubber.add_response(
+            method='get_object',
+            service_response={
+                'ContentLength': len(b_contents),
+                'Body': BytesIO(b_contents)
+            },
+            expected_params={
+                'Bucket': 'test-bucket',
+                'Key': 'dir/foo/b'
+            }
+        )
+
+        with patch('quilt3.data_transfer.s3_transfer_config.max_request_concurrency', 1):
+            bucket.fetch('dir/', './')
+
+        assert pathlib.Path('a').read_bytes() == a_contents
+        assert pathlib.Path('foo/b').read_bytes() == b_contents
+
+        # Fetch a single file.
+
+        self.s3_stubber.add_response(
+            method='head_object',
```

Comment:
where does this response actually get consumed?

Reply:
Before downloading the file, we "head" it to get the size and decide whether to do a multipart download. That doesn't happen in the directory download because listing the directory gives us sizes.

(That's also what AWS SDK does.)

Reply:
that's what i figured, but then above i saw a `.fetch` call with no `head_object` on https://github.com/quiltdata/quilt/pull/1467/files#diff-e2e375e7182ff0261068d685af014694R65